File: iosApp/iosApp/Modules/CurrencySwap ✅/CurrencySwapWireframeFactory.swift
// Created by web3d4v on 14/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

protocol CurrencySwapWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: CurrencySwapWireframeContext
    ) -> CurrencySwapWireframe
}

final class DefaultCurrencySwapWireframeFactory {
    private let currencyPickerWireframeFactory: CurrencyPickerWireframeFactory
    private let confirmationWireframeFactory: ConfirmationWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let walletService: WalletService
    private let networksService: NetworksService
    private let swapService: UniswapService
    private let currencyStoreService: CurrencyStoreService

    init(
        currencyPickerWireframeFactory: CurrencyPickerWireframeFactory,
        confirmationWireframeFactory: ConfirmationWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        walletService: WalletService,
        networksService: NetworksService,
        swapService: UniswapService,
        currencyStoreService: CurrencyStoreService
    ) {
        self.currencyPickerWireframeFactory = currencyPickerWireframeFactory
        self.confirmationWireframeFactory = confirmationWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.walletService = walletService
        self.networksService = networksService
        self.swapService = swapService
        self.currencyStoreService = currencyStoreService
    }
}

extension DefaultCurrencySwapWireframeFactory: CurrencySwapWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: CurrencySwapWireframeContext
    ) -> CurrencySwapWireframe {
        
        DefaultCurrencySwapWireframe(
            parent,
            context: context,
            currencyPickerWireframeFactory: currencyPickerWireframeFactory,
            confirmationWireframeFactory: confirmationWireframeFactory,
            alertWireframeFactory: alertWireframeFactory,
            walletService: walletService,
            networksService: networksService,
            swapService: swapService,
            currencyStoreService: currencyStoreService
        )
    }
}

// MARK: - Assembler

final class CurrencySwapWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> CurrencySwapWireframeFactory in
            DefaultCurrencySwapWireframeFactory(
                currencyPickerWireframeFactory: resolver.resolve(),
                confirmationWireframeFactory: resolver.resolve(),
                alertWireframeFactory: resolver.resolve(),
                walletService: resolver.resolve(),
                networksService: resolver.resolve(),
                swapService: resolver.resolve(),
                currencyStoreService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/currencyswap/CurrencySwapWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.currencyswap

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.uniswap.UniswapService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapWireframe
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapWireframeContext
import smartadapter.internal.extension.name

interface CurrencySwapWireframeFactory {
    fun make(fragment: Fragment?, context: CurrencySwapWireframeContext): CurrencySwapWireframe
}

class DefaultCurrencySwapWireframeFactory(
    private val walletService: WalletService,
    private val networksService: NetworksService,
    private val swapService: UniswapService,
    private val currencyStoreService: CurrencyStoreService,
): CurrencySwapWireframeFactory {

    override fun make(
        fragment: Fragment?,
        context: CurrencySwapWireframeContext
    ): CurrencySwapWireframe = DefaultCurrencySwapWireframe(
        fragment?.let { WeakRef(it) },
        context,
        walletService,
        networksService,
        swapService,
        currencyStoreService,
    )
}

class CurrencySwapWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(CurrencySwapWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultCurrencySwapWireframeFactory(
                it.resolve(WalletService::class.name),
                it.resolve(NetworksService::class.name),
                it.resolve(UniswapService::class.name),
                it.resolve(CurrencyStoreService::class.name),
            )
        }
    }
}